src/features/left-sidebar/LeftSidebarDropWrapper.js

```diff
@@ -9,24 +10,63 @@ type Props = {
     dropTargetRef?: { current: null | HTMLDivElement },
     isDragging?: boolean,
     message?: string,
+    showDropZoneOnHover?: boolean,
 };
 
-const LeftSidebarDropWrapper = ({
-    children,
-    className = '',
-    isDragging = false,
-    message = '',
-    dropTargetRef,
-    ...rest
-}: Props) => (
-    <div ref={dropTargetRef} className={`left-sidebar-drop-wrapper ${className}`} {...rest}>
-        {isDragging ? (
-            <div className="left-sidebar-drop-veil">
-                <span className="left-sidebar-drop-wrapper-text">{message}</span>
+type State = {
+    dropZoneHover: boolean,
+};
+
+class LeftSidebarDropWrapper extends React.Component<Props, State> {
+    state = {
+        dropZoneHover: false,
+    };
+
+    componentDidUpdate(prevProps: Props) {
+        const { showDropZoneOnHover, isDragging } = this.props;
+
+        // Reset drop zone hover state if dragging has stopped without firing a mouseleave event.
+        if (showDropZoneOnHover && !isDragging && prevProps.isDragging) {
+            this.setState({ dropZoneHover: false });
+        }
+    }
+
+    handleDropZoneHover = () => this.setState({ dropZoneHover: true });
+
+    handleDropZoneHoverLeave = () => this.setState({ dropZoneHover: false });
+
+    render() {
+        const {
+            children,
+            className = '',
+            isDragging = false,
+            showDropZoneOnHover = false,
+            message = '',
+            dropTargetRef,
+            ...rest
+        } = this.props;
+        const { dropZoneHover } = this.state;
+        const shouldShowVeil = isDragging && (showDropZoneOnHover ? dropZoneHover : true);
+        const hoverEventHandlers =
+            isDragging && showDropZoneOnHover
+                ? {
+                      onMouseEnter: this.handleDropZoneHover,
+                      onMouseLeave: this.handleDropZoneHoverLeave,
+                  }
+                : {};
+        const classes = classNames('left-sidebar-drop-wrapper', className);
+
+        return (
+            <div ref={dropTargetRef} className={classes} {...hoverEventHandlers} {...rest}>
+                {shouldShowVeil ? (
+                    <div className="left-sidebar-drop-veil">
+                        <span className="left-sidebar-drop-wrapper-text">{message}</span>
+                    </div>
+                ) : null}
```

Comment:
Minor: you could shorten this logic by replacing the ternary statement with `&&`.